Capital locations

mchenryc opened the pull request:
fixes #112

mchenryc committed:
performance fix - remove unnecessary loop

mchenryc committed:
format freeciv-web/README.md

mchenryc committed:
Report opponent capital in intel-report if known

mchenryc committed:
Add additional improvement constant

lonemadmax commented on freeciv-web/src/main/webapp/javascript/player.js:
```diff
@@ -247,3 +247,20 @@ function player_has_wonder(playerno, improvement_id)
+function player_capital(player)
+{
+  return Object.values(cities)
+    .find(city => does_player_own_city(player, city) && is_capital(city));
```
Can't use Object.values nor Array.find. Our friend MSIE again(*), and this time they are not translated in compilation.
This one is still readable even if more verbose in old JS. Or maybe we should start adding polyfills and forget these things once and for all.

(*) No, I haven't tested. Does one of the libraries already provide a definition for those functions?

mchenryc replied:
Something _does_ polyfill it\*, but I'm still searching for what might be doing so.

I've reverted to the `for...in` construct in common use, leaving the polyfill question for other issues.

\* From an IE console running freeciv-web `Object.values` is defined and works as expected, while not so in a default window console.

lonemadmax commented on freeciv-web/src/main/webapp/javascript/improvement.js:
```diff
@@ -19,6 +19,9 @@
 
+// incomplete list of (well defined) building IDs - populate as needed
+const B_AIRPORT = 0;
+const B_PALACE = 21;
```
That's 23 with civ2civ3 ruleset.

mchenryc replied:
Dynamic? Good catch - back to the drawing board.

FWIW, I find our PR back-and-fourth very useful in understanding "edge" (sometimes not so edge...) cases which need to be checked. Thanks again, for your help and diligence.

I expect it has come up, but is there a discussion going on anywhere around automating a test suite and how to set up a framework for testing various edge cases - something more targeted than just running an autogame for N turns?

lonemadmax replied:
Not that I knew it depended on ruleset. I found out when looking for the source of that 21.

We had a similar problem with unit orders that depend on technologies. As a "temporary" fix, the ids were changed to a search by name (https://github.com/freeciv/freeciv-web/commit/1e3ec23ade0b6038ca7a1149381a22aa36b4b789, https://github.com/freeciv/freeciv-web/commit/60ad3e448d7eb84eeffd46ff91025a3140f1ca1d). The same search should work here with the `improvements` list.

Re test, I'm not aware of any discussion, neither automatic nor manual, not even a checklist. I guess it could be done, with a large sample of crafted savegames and casper/selenium/whatever scripts to check values and the results of different actions. It would also be a hard, thankless job.

mchenryc replied:
I suspect the dynamic `EXTRA_*` (858cc4c6) variables I encountered in (ba71af03) are related. Any recommendations on how to get up-to speed on Effects is welcome - I see [effects.*](https://github.com/freeciv/freeciv/blob/27a777a32ef0faa0bcca819c10f2fab4dcf15d51/common/effects.c) and other refernces, but no prose description in the [HACKING](https://github.com/freeciv/freeciv/blob/master/doc/HACKING) file - I find discussions of intent and descriptions of design decisions very useful guides before diving into what I like to call "code archaeology". I'll eventually end up looking in forums or mailing lists, but if you know of discussions, it would be great to mention them in #208.

mchenryc replied:
fixed (see commit message).

mchenryc committed:
Create improvements index, and guard write access

Write access is not actually 'guarded', but writing is now only done by
methods in improvements.js, making the file, and data it manipulates,
more modular.

This reuduces complexity, allowing improvements.js to be sole owner of
data structure it declares.

Indeed, a second data structure (improvements_name_index) is declared to
improve runtime lookup. As variables are still global, longer names are
still in use to prevent clashes.

Read access is still through the global 'improvements' variable, which
could still cause issues.

lonemadmax commented on freeciv-web/src/main/webapp/javascript/player.js:
```diff
@@ -271,3 +271,25 @@ function get_invalid_username_reason(username)
+      return true;
+    }
+  }
+  return false;
```
You surely mean to return `city` and `undefined`.
Sorry I didn't spot this before.

mchenryc replied:
Indeed, though you didn't miss it: this originally used `Object.values(...)`.

mchenryc replied:
Also adjusted improvement_id_by_name to always return an int value - I realized that if a ruleset omitted a requested building, returning `undefined` or `null` could in some cases be considered the equivalent to `0` given js type-conversion.
